pkg/channel/attributes/dispatch_error.go

```diff
@@ -0,0 +1,61 @@
+/*
+Copyright 2020 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package attributes
+
+import (
+	"encoding/json"
+
+	"github.com/cloudevents/sdk-go/v2/binding"
+	"github.com/cloudevents/sdk-go/v2/binding/transformer"
+)
+
+const (
+	DispatchErrorExtensionKey          = "knativedispatcherr" // Max 20 characters from "a-z0-9" starting with "knative"
+	MaxDispatchErrorExtensionDataBytes = 1024
+)
+
+// The DispatchError Extension Value Struct
+type DispatchErrorExtension struct {
+	Code int    `json:"code,omitempty"`
+	Data []byte `json:"data,omitempty"`
+}
+
+// DispatchErrorExtension Constructor
+func NewDispatchErrorExtension(code int, data []byte) DispatchErrorExtension {
+	return DispatchErrorExtension{Code: code, Data: data}
+}
+
+// Returns An AddExtension Transformer For The DispatchErrorExtension Attribute
+func (d *DispatchErrorExtension) AddExtensionTransformer() (binding.TransformerFunc, error) {
+
+	// The DispatchErrorExtension To Add As An Extension
+	dispatchErrorExtension := *d
+
+	// Trim Data Field To Max Length If Necessary (Immutable Copy)
+	if len(d.Data) > MaxDispatchErrorExtensionDataBytes {
+		dispatchErrorExtension = NewDispatchErrorExtension(d.Code, d.Data[:MaxDispatchErrorExtensionDataBytes])
+	}
+
+	// Marshal The DispatchErrorExtension Into JSON Bytes
+	jsonBytes, err := json.Marshal(dispatchErrorExtension)
+	if err != nil {
+		return nil, err
+	}
+
+	// Create & Return A CloudEvent AddExtension Transformer
+	return transformer.AddExtension(DispatchErrorExtensionKey, jsonBytes), nil
+}
```

Comment:
TBH I prefer that the error extension is splitted into two extension attributes, more than wrapping them in a json and smashing it in a single ext attribute. Even if the spec allows such things, It's generally a bad idea becasue it doesn't play nicely with "binary mode" of protocols like http. Also some sdks doesn't really support it properly.

Look for example at this sequence extension: https://github.com/cloudevents/spec/blob/master/extensions/sequence.md It's splitted in two extension attributes, although "logically" it's the same extension.

What about splitting to `knativeerrorcode` and `knativeerrordata`? It also sounds "generic enough" to be used in other contexts in future.

Reply:
Sure, I can split it into two extension attributes as suggested.  The type of the `knativeerrordata` will continue to be the raw []byte from the HTTP Response.Body or would you rather it be a string?

Reply:
Also, now that the extensions are more "generic" and not "dispatch" related is the current location in `pkg/channel/attributes` still acceptable or is there a better place for them?  Maybe just up one level at `pkg/attributes` ?

Reply:
>  The type of the knativeerrordata will continue to be the raw []byte from the HTTP Response.Body or would you rather it be a string?

I think it should be string, usually an error is in a string format, not binary. AFAIR providing to `SetExtension` a `[]byte` serializes it to base64, I think there's no reason to do that, while with string format people can easily diagnose it. 

> Also, now that the extensions are more "generic" and not "dispatch" related is the current location in pkg/channel/attributes still acceptable or is there a better place for them?

Let's keep it there, and we can move it in future when we want to use them somewhere else